### PR TITLE
LibWeb: Implement WebCrypto HKDF

### PR DESCRIPTION
**Pull request by BenWiederhake**
This PR:
- Implements and tests basic HKDF functionality
- Implements the bindings of HKDF for SubtleCrypto
- Add a test in LibWeb, which also covers the weird edge case that while RFC 5869 [defines in Section 2.2](https://www.rfc-editor.org/rfc/rfc5869#section-2.2) that explicitly provides the feature "optional salt value \[…\] if not provided, it is set to \[…\]", neither the spec nor other browsers pass this feature through to the JS callers. In other words, JS *requires* the salt value, even though RFC 5869 says it's optional.

As a result, we now pass hundreds of additional WPT tests.
- We used to be at 1/1001: [FIXME](https://wpt.fyi/results/WebCryptoAPI/derive_bits_keys/hkdf.https.any.html%3F1-1000?sha=100f9048d8&product=ladybird)
- And now we're apparently at 343/1001:
Bildschirmfoto_2024-10-20_06-54-39

Note that similar WPT failures regarding HMAC could be fixed trivially, but I want to keep this PR small for now.

### CODE REVIEW

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp [hunk 1]
```diff
@@ -939,6 +965,51 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<JS::Object>> RSAOAEP::export_key(Bindings::
     return JS::NonnullGCPtr { *result };
 }
 
+WebIDL::ExceptionOr<JS::NonnullGCPtr<CryptoKey>> HKDF::import_key(AlgorithmParams const&, Bindings::KeyFormat format, CryptoKey::InternalKeyData key_data, bool extractable, Vector<Bindings::KeyUsage> const& key_usages)
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:968]: Could we also get a link to the spec?

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp [hunk 2]
```diff
@@ -1383,6 +1454,67 @@ WebIDL::ExceptionOr<JS::Value> ED25519::verify([[maybe_unused]] AlgorithmParams
...
+    auto const& normalized_algorithm = static_cast<HKDFParams const&>(params);
+
+    // 1. If length is null or zero, or is not a multiple of 8, then throw an OperationError.
+    auto length = length_optional.value_or(0);
+
+    if (length == 0 || length % 8 != 0)
+        return WebIDL::OperationError::create(realm, "Length must be greater than 0 and divisible by 8"_string);
+
+    // 2. Let extractKey be a key equal to n zero bits where n is the size of the output of the hash function described by the hash member of normalizedAlgorithm.
+    // (However, this variable is never directly used, and therefore pointless.)
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:1470]: Maybe spec bug report?

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp [hunk 3]
```diff
@@ -1383,6 +1454,67 @@ WebIDL::ExceptionOr<JS::Value> ED25519::verify([[maybe_unused]] AlgorithmParams
...
+
+    // 3. Let keyDerivationKey be the secret represented by [[handle]] internal slot of key as the message.
+    auto keyDerivationKey = key->handle().visit(
+        [](ByteBuffer data) -> ByteBuffer {
+            return data;
+        },
+        [](auto) -> ByteBuffer { VERIFY_NOT_REACHED(); });
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:1477]: Eh, why not `.get<ByteBuffer>()` instead of this?

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp [hunk 4]
```diff
@@ -1383,6 +1454,67 @@ WebIDL::ExceptionOr<JS::Value> ED25519::verify([[maybe_unused]] AlgorithmParams
...
+    auto length = length_optional.value_or(0);
+
+    if (length == 0 || length % 8 != 0)
+        return WebIDL::OperationError::create(realm, "Length must be greater than 0 and divisible by 8"_string);
+
+    // 2. Let extractKey be a key equal to n zero bits where n is the size of the output of the hash function described by the hash member of normalizedAlgorithm.
+    // (However, this variable is never directly used, and therefore pointless.)
+
+    // 3. Let keyDerivationKey be the secret represented by [[handle]] internal slot of key as the message.
+    auto keyDerivationKey = key->handle().visit(
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:1473]: `key_derivation_key` please.

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp [hunk 5]
```diff
@@ -1383,6 +1454,67 @@ WebIDL::ExceptionOr<JS::Value> ED25519::verify([[maybe_unused]] AlgorithmParams
...
+        },
+        [](auto) -> ByteBuffer { VERIFY_NOT_REACHED(); });
+
+    // 4. Let result be the result of performing the HKDF extract and then the HKDF expand step described in Section 2 of [RFC5869] using:
+    //    * the hash member of normalizedAlgorithm as Hash,
+    //    * keyDerivationKey as the input keying material, IKM,
+    //    * the contents of the salt member of normalizedAlgorithm as salt,
+    //    * the contents of the info member of normalizedAlgorithm as info,
+    //    * length divided by 8 as the value of L,
+    // FIXME: salt null versus salt empty?!
```
Comment (alimpfard) [inline Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp]: Maybe spec bug report 2?

**Review by alimpfard — COMMENTED** [review]
Code looks ok (mostly style-wise), but looks like there's some pointer mess-up somewhere (see ASAN output)?

Also you can also use HashManager for the big if-elae chain to avoid repeating the HKDF call if you want.

File: Tests/LibCrypto/TestHKDF.cpp [hunk 6]
```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2023, stelar7 <dudedbz@gmail.com>
```
Comment (stelar7) [inline Tests/LibCrypto/TestHKDF.cpp:2]: not me doing this 😅

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp (hunk 2 above)
Comment (BenWiederhake) [reply Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:1470]: It's not exactly a *bug*, it's merely a missing feature that (IMHO) shouldn't exist anyway. Feel free to report it, but I'm not going to.

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp (hunk 3 above)
Comment (BenWiederhake) [reply Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:1477]: I "inspired" myself by the PBKDF2-implementation a few lines down. I'll change it in a new commit, thank you!

File: Tests/LibCrypto/TestHKDF.cpp (hunk 6 above)
Comment (BenWiederhake) [reply Tests/LibCrypto/TestHKDF.cpp:2]: Initially I borrowed a lot of code from TestPBKDF2, which you did write. However, by now barely anything remains ^^'

I'll leave you in, but also add myself (no idea how I forgot that part).

**Comment by BenWiederhake** [comment]
> Code looks ok (mostly style-wise), but looks like there's some pointer mess-up somewhere (see ASAN output)?

Do you mean this error?

```
/home/runner/work/ladybird/ladybird/AK/Span.h:190:26: runtime error: null pointer passed as argument 2, which is declared to never be null
```

It looks like this is a bug in the HMAC code itself, since it assumes that the key-buffer is non-empty. I'll fix that in a separate PR.

File: Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp [hunk 7]
```diff
@@ -1383,6 +1454,67 @@ WebIDL::ExceptionOr<JS::Value> ED25519::verify([[maybe_unused]] AlgorithmParams
...
+    auto const& hash_algorithm = TRY(normalized_algorithm.hash.visit(
+        [](String const& name) -> JS::ThrowCompletionOr<String> { return name; },
+        [&](JS::Handle<JS::Object> const& obj) -> JS::ThrowCompletionOr<String> {
+            auto name_property = TRY(obj->get("name"));
+            return name_property.to_string(m_realm->vm()); }));
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/Crypto/CryptoAlgorithms.cpp:1493]: I swear I asked stelar7 or someone else to add a helper for this pattern 🤔 . Did some of those PRs get lost in the fork?

File: Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+        // Test the subtle difference between an empty salt (0 bytes) and an absent salt (which defaults to hashLen many bytes, each of value 0).
+        // Inspired by https://datatracker.ietf.org/doc/html/rfc5869/#appendix-A.6 and A.7
+        var algorithmEmptySalt = {name: "HKDF", salt: new Uint8Array([]), info: new Uint8Array([]), hash: "SHA-1"};
+        var bitsEmptySalt = await subtle.deriveBits(algorithmEmptySalt, key, 42 * 8);
+        console.log(new Uint8Array(bitsEmptySalt));
+        println(new Uint8Array(bitsEmptySalt));
+        var algorithmAbsentSalt = {name: "HKDF", salt: null, info: new Uint8Array([]), hash: "SHA-1"};
+        subtle.deriveBits(algorithmAbsentSalt, key, 42 * 8).then(
+            async (bitsAbsentSalt) => { println(new Uint8Array(bitsAbsentSalt)); done(); },
+            async (err) => { println("Absent salt rejected! " + err); done(); }
```
Comment (ADKaster) [inline Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html:16]: I'm skeptical that prettier would format this code like this. Though we don't run it on html files in CI at the moment.

Comment (BenWiederhake) [reply Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html:16]: Thanks! I installed prettier and reformatted it.

**Comment by BenWiederhake** [comment]
Changes since last push:
- Mention importkey spec url (thanks, @alimpfard!)
- Fix HMAC nonnull problem (thanks, @alimpfard and CI!)
- Ran prettier on new JS test code (thanks, @ADKaster!)
- Use ".get<ByteBuffer>()" to simplify access (thanks, @alimpfard!)
- Propagate ".get<ByteBuffer>()" simplification to entire file. There seem to be no other occurrences of this pattern in Ladybird.
- Fix "keyDerivationKey" naming (thanks, @alimpfard!)
- Updated TestHKDF copyright (thanks, @stelar7!)
- Rebased on master for convenience

The HashManager idea is good, but cannot be used directly, so I skipped that part.

The convenience function for the "hash object to hash name conversion" doesn't appear to have already been merged, so I cannot use it just yet.

Also, I opened a spec issue; let's see what comes from that. No action required from us, though.

**Comment by stelar7** [comment]
> The convenience function for the "hash object to hash name conversion" doesn't appear to have already been merged, so I cannot use it just yet.

Feel free to steal the commit from here: https://github.com/LadybirdBrowser/ladybird/commit/fb9d35a0ca451b42cd79e16bae445d74fb096a5a

**Comment by BenWiederhake** [comment]
Thanks, but neither did I write the code, nor do I fully understand it. Let's treat that as a separate PR :)

**Review by alimpfard — APPROVED** [review]

**Review by ADKaster — APPROVED** [review]

File: Userland/Libraries/LibCrypto/Hash/HKDF.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            salt_buffer.resize(DigestType::Size, ByteBuffer::ZeroFillNewElements::Yes);
+            return salt_buffer.bytes();
+        });
+        HMACType hmac_salt(salt);
+
+        // https://www.rfc-editor.org/rfc/rfc5869#section-2.2
+        // PRK = HMAC-Hash(salt, IKM)
+        auto prk_digest = hmac_salt.process(input_keying_material);
+        auto prk = prk_digest.bytes();
+        ASSERT(prk.size() == DigestType::Size);
```
Comment (nico) [inline Userland/Libraries/LibCrypto/Hash/HKDF.h:47]: Did you intentionally use ASSERT here (and elsewhere)? From what I understand, most places use VERIFY(), and ASSERT() is for performance-critical code.

File: Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        // Test the subtle difference between an empty salt (0 bytes) and an absent salt (which defaults to hashLen many bytes, each of value 0).
+        // Inspired by https://datatracker.ietf.org/doc/html/rfc5869/#appendix-A.6 and A.7
+        var algorithmEmptySalt = {
+            name: "HKDF",
+            salt: new Uint8Array([]),
+            info: new Uint8Array([]),
+            hash: "SHA-1",
+        };
+        var bitsEmptySalt = await subtle.deriveBits(algorithmEmptySalt, key, 42 * 8);
+        console.log(new Uint8Array(bitsEmptySalt));
```
Comment (nico) [inline Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html:19]: Left-over debug logging?

File: Userland/Libraries/LibCrypto/Hash/HKDF.h (hunk 9 above)
Comment (BenWiederhake) [reply Userland/Libraries/LibCrypto/Hash/HKDF.h:47]: Good catch! No, that was a mistake. Now that I know the difference (thanks for the explanation!), this should be `VERIFY` instead.

File: Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html (hunk 10 above)
Comment (BenWiederhake) [reply Tests/LibWeb/Text/input/subtle-crypto-hkdf-salt-empty-or-none.html:19]: Oh no! Yes, this is left-over debug logging.
